algebra/ring.lean

```diff
@@ -20,77 +20,64 @@ section units
   variable (α)
 
   structure invertible :=
-    (val : α)
-    (inv : α)
-    (property : val * inv = 1 ∧ inv * val = 1)
+  (val : α)
+  (inv : α)
+  (val_inv : val * inv = 1)
+  (inv_val : inv * val = 1)
 
   variables {a b c : invertible α}
   variable {α}
 
-  def ext (HAB : a.val = b.val) : a = b :=
+  def units.ext (HAB : a.val = b.val) : a = b :=
   begin
     cases a,
     cases b,
     dsimp at HAB,
     congr,
     exact HAB,
     exact calc
-        inv = inv * (val_1 * inv_1) : by rw [property_1.1, mul_one]
+        inv = inv * (val_1 * inv_1) : by rw [val_inv_1, mul_one]
         ... = (inv * val_1) * inv_1 : by rw [mul_assoc]
         ... = (inv * val) * inv_1   : by rw [HAB]
-        ... = inv_1                 : by rw [property.2, one_mul]
+        ... = inv_1                 : by rw [inv_val, one_mul]
   end
 
   lemma mul_four {a b c d : α} :
   (a * b) * (c * d) = a * (b * c) * d := by simp
 
-  def mul (a b : invertible α) : invertible α :=
-  {
-    val := a.val * b.val,
-    inv := b.inv * a.inv,
-    property := ⟨by rw [mul_four,b.property.1,mul_one,a.property.1],
-                by rw [mul_four,a.property.2,mul_one,b.property.2]⟩
-  }
-
-  def one : invertible α :=
-  {
-    val := 1,
-    inv := 1,
-    property := ⟨mul_one 1, one_mul 1⟩
-  }
-
-  def inv (a : invertible α) : invertible α :=
-  {
-    val := a.inv,
-    inv := a.val,
-    property := ⟨a.property.2, a.property.1⟩
-  }
-
-  instance : has_mul (invertible α) := ⟨mul⟩
-  instance : has_one (invertible α) := ⟨one⟩
-  instance : has_inv (invertible α) := ⟨inv⟩
+  instance : has_mul (invertible α) := ⟨(λ a b,
+  { val     := a.val * b.val,
+    inv     := b.inv * a.inv,
+    val_inv := by rw [mul_four,b.val_inv,mul_one,a.val_inv],
+    inv_val := by rw [mul_four,a.inv_val,mul_one,b.inv_val] })⟩
```

Comment:
It's a matter of taste, but I'd prefer to use anonymous constructors for this (it's not a class, and it's being treated like data). See how functions are defined on `rat` for inspiration.

Also, I prefer to destruct the arguments in the initial match, rather than using the projections. Unfortunately that means using a separate definition `protected def mul` in this case, but we can still avoid references to it later on.